# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [1]:
# TODO: import the pandas library
import pandas as pd
# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4)
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4)

# TODO: remove the 'Unnamed: 62' column from each data set
df_rural.drop('Unnamed: 62', axis=1, inplace=True)
df_electricity.drop('Unnamed: 62', axis=1, inplace=True)
# TODO: combine the two data sets together using the concat method

In [2]:
df_rural.head(2)

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   

   Indicator Code    1960    1961    1962   1963    1964   1965  ...    2008  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.27  49.285  49.30  ...  56.217   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.89  90.574  90.25  ...  76.054   

     2009    2010    2011    2012    2013    2014    2015    2016    2017  
0  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696  58.893  
1  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868  72.430  

[2 rows x 62 columns]

In [3]:
df_rural.shape

(264, 62)

In [4]:
df_electricity.head(2)

Country Name Country Code                           Indicator Name  \
0        Aruba          ABW  Access to electricity (% of population)   
1  Afghanistan          AFG  Access to electricity (% of population)   

   Indicator Code  1960  1961  1962  1963  1964  1965  ...       2008  \
0  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  93.086166   
1  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  42.400000   

        2009       2010       2011       2012       2013       2014  \
0  93.354546  93.356292  93.942375  94.255814  94.578262  94.906723   
1  44.854885  42.700000  43.222019  69.100000  67.259552  89.500000   

        2015       2016  2017  
0  95.238182  95.570145   NaN  
1  71.500000  84.137138   NaN  

[2 rows x 62 columns]

In [5]:
df_electricity.shape

(264, 62)

In [6]:
df = pd.concat([df_rural, df_electricity], axis=0)
df.shape

(528, 62)

# Exercise 2 (Challenge)

This exercise is more challenging.

The resulting data frame should look like this:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [7]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural_melted = pd.melt(df_rural, id_vars=['Country Name', 'Country Code'], value_vars=(df_rural.columns.drop(
    ['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code']).values), var_name='Year', 
                          value_name='Rural_Value')

df_electricity_melted = pd.melt(df_electricity, id_vars=['Country Name', 'Country Code'], value_vars=(
    df_electricity.columns.drop(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code']).values), var_name='Year',
                               value_name='Electricity_Value')
# TODO: drop any columns from the data frames that aren't needed

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year

# TODO: sort the results by country and then by year

print(df_rural_melted.head(), '\n', df_electricity_melted.head())

  Country Name Country Code  Year  Rural_Value
0        Aruba          ABW  1960       49.224
1  Afghanistan          AFG  1960       91.779
2       Angola          AGO  1960       89.565
3      Albania          ALB  1960       69.295
4      Andorra          AND  1960       41.550 
   Country Name Country Code  Year  Electricity_Value
0        Aruba          ABW  1960                NaN
1  Afghanistan          AFG  1960                NaN
2       Angola          AGO  1960                NaN
3      Albania          ALB  1960                NaN
4      Andorra          AND  1960                NaN


In [8]:
df_combined = df_rural_melted.merge(df_electricity_melted, how='outer', on=['Country Name', 'Country Code', 'Year'])
df_combined.sort_values(by=['Country Name', 'Year'], ascending=True, inplace=True, ignore_index=True)

df_combined.head()

Country Name Country Code  Year  Rural_Value  Electricity_Value
0  Afghanistan          AFG  1960       91.779                NaN
1  Afghanistan          AFG  1961       91.492                NaN
2  Afghanistan          AFG  1962       91.195                NaN
3  Afghanistan          AFG  1963       90.890                NaN
4  Afghanistan          AFG  1964       90.574                NaN